Add `cirq.convert_to_target_gateset` transformer and `cirq.CompilationTargetGateset` interface

tanujkhattar opened the pull request:
- Adds the following new objects:
  - `cirq.CompilationTargetGateset`: Interface for defining compilation targets. 
  - `cirq.decompose_operations_to_target_gateset` : Transformer to locally convert each operation to target gateset.
  - `cirq.convert_to_target_gateset`: Transformer to convert a circuit to a target gateset -- also supports running pre and post processing transformers before decomposing individual operations.
- See https://github.com/quantumlib/Cirq/pull/5003 for larger context and implementations of specific target gatesets deriving from `cirq.CompilationTargetGateset`. 
---- 

- Part of https://github.com/quantumlib/Cirq/issues/4722 
- Follows the new Transformer API https://github.com/quantumlib/Cirq/issues/4483
- Supports no compile tags NoCompile Tag for optimizers https://github.com/quantumlib/Cirq/issues/4253


cc @MichaelBroughton @dstrain115

MichaelBroughton commented on cirq-core/cirq/transformers/convert_to_target_gateset.py:
```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Nit: module docstring.

MichaelBroughton commented on cirq-core/cirq/transformers/convert_to_target_gateset.py:
```diff
@@ -0,0 +1,129 @@
+
+from typing import Optional, Callable, TYPE_CHECKING
+
+from cirq import protocols
+from cirq.transformers import transformer_api, transformer_primitives
+from cirq.protocols.decompose_protocol import DecomposeResult
```
Nit: module imports.

tanujkhattar replied:
Done

MichaelBroughton commented on cirq-core/cirq/transformers/convert_to_target_gateset.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@transformer_api.transformer
+def decompose_operations_to_target_gateset(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    gateset: Optional['cirq.Gateset'] = None,
+    decomposer: Callable[['cirq.Operation', int], DecomposeResult] = lambda *_: NotImplemented,
+    ignore_failures=True,
+) -> 'cirq.Circuit':
```
Do we need this transformer ? It seems like a user should be able to fairly easily accomplish this same thing by just calling into the regular `decompose`. Having this as a thin wrapper over decompose seems like it might be complicating things for the users.

tanujkhattar replied:
Replied in the main comment. Tl;Dr is that having benefits of making it a transformer, like automated logging + no-compile tags, make it worth it. Plus we have other similar wrappers like `expand_composite`.

MichaelBroughton commented on cirq-core/cirq/transformers/convert_to_target_gateset_test.py:
```diff
@@ -0,0 +1,200 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: skip-file
```
Do we want to skip the whole file ?

tanujkhattar replied:
Changed to only disable line too long on a single test.

MichaelBroughton commented on cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py:
```diff
@@ -0,0 +1,98 @@
+    which can transform any given circuit to contain gates accepted by this gateset.
+    """
+
+    @property
+    @abc.abstractmethod
+    def num_qubits(self) -> int:
+        """Maximum number of qubits on which a gate from this gateset can act upon."""
```
Can this be defined here as something like:
```python3
return max(g.num_qubits() for g in self.gates)
```
 ?

tanujkhattar replied:
Gateset is composed of gate families, and gate families don't have a `num_qubits` parameter. For example, it's possible to create gate families which can accept any controlled single qubit rotation with one or more controls. A `num_qubits` parameter cannot be precisely defined in this case. 

Enforcing `num_qubits` on gate families would be a much broader change without significant benefit outside the scope of compilation target gatesets.

MichaelBroughton commented on cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py:
```diff
@@ -0,0 +1,98 @@
+
+
+class CompilationTargetGateset(ops.Gateset, metaclass=abc.ABCMeta):
+    """Abstract base class to create gatesets that can be used as targets for compilation.
+
+    An instance of this type can be passed to transformers like `cirq.convert_to_target_gateset`,
+    which can transform any given circuit to contain gates accepted by this gateset.
```
This class seems important enough for users that it might be good to give a snippet of a simple implementation in the docstring. WDYT ?

tanujkhattar replied:
Concrete implementations will be added in the follow-up PR. Maybe we can then update the docstrings here to include a hyper-linked example to implementation? Something like 

"Please see `cirq.CZTargetGateset` for a concrete implementation." ?

MichaelBroughton reviewed (commented):
Few small nits and then a few high level questions. My biggest concern is if we should ship something in `decompose_operations_to_target_gateset` that looks so similar to `decompose`. My thinking is that less is more and it's probably best if we don't give users too many ways to do the same thing.

tanujkhattar commented:
> My biggest concern is if we should ship something in decompose_operations_to_target_gateset that looks so similar to decompose.


The primary benefit here is the automated logging + support for no-compile tags, that comes with making it a transformer. 
For example: If someone runs `convert_to_target_gateset` with a context and then calls `context.logger.show()`, they'd be able to clearly inspect the action of each stage, including the decomposition stage wrapped in `decompose_operations_to_target_gateset`, because of the automated logging framework. 


Given this benefit, and given that we already have other transformers like `expand_composite`, which are essentially wrappers on top of `decompose` + `map_operations`, I'd argue that we should keep this `decompose_operations_to_target_gateset` as well.

MichaelBroughton commented on cirq-core/cirq/transformers/convert_to_target_gateset.py:
```diff
@@ -0,0 +1,130 @@
+    context: Optional['cirq.TransformerContext'] = None,
+    gateset: Optional['cirq.Gateset'] = None,
+    decomposer: Callable[['cirq.Operation', int], dp.DecomposeResult] = lambda *_: NotImplemented,
+    ignore_failures=True,
```
Nit: type

MichaelBroughton reviewed (approved):
LGTM. The closeness of `decompose_operations_to_target_gateset` to `decompose` still worries me a bit. Could we do something like:

`decompose_operationst_to_target_gateset` -> `convert_to_gateset`
`convert_to_target_gateset` -> `fuse_and_convert_to_gateset`

Just to avoid the overlap with decompose and in the case of `fuse_and_convert_to_gateset` communicate the fact that it won't be just a simple conversion. WDYT ?

tanujkhattar commented:
As discussed offline, 

- Changed `decompose_operations_to_target_gateset` --> `_decompose_operations_to_target_gateset` to avoid polluting the global namespace.
- Changed `convert_to_target_gateset` --> `optimize_for_target_gateset` to convey that other pre/post process transformers will also be executed.
